Fixed 'Delete operation is forbidden for current area' error.

vovayatsyuk opened the pull request:
### Description
This error happens if the customer address doesn't pass validation during account creation.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#9736
2. Fixes magento/magento2#8631
3. Fixes magento/magento2#5699

### Manual testing scenarios
1. Open Magento/Customer/Model/AccountManagement.php and find the following lines:

```php
        try {
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

2. Change them as shown below (Emulate InputException from addressRepository->save method):

```php
        try {
            throw new InputException(__("Test"));                     // <-- this line was added
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

3. Add product to cart as a guest and navigate to multishipping checkout.
4. Press "Create account"
5. Fill the form and submit it.
6. An error "Test" will be shown.
7. If you will try to register one more time, you'll be informed that this email is already taken.
8. You can also check the customer's grid in the magento backend and see that your customer was just created.

### Expected Behavior

The client should see the "Test" error message every time he submits the form.

### Actual Behavior

The client sees the "Delete operation is forbidden for current area" error message for first time. Next submits tells that his email is already in use.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#31540: Fixed 'Delete operation is forbidden for current area' error.

vovayatsyuk committed:
Fixed 'Delete operation is forbidden for current area' error.
This error happens if customer address doesn't pass validation during account creation.

rogyar reviewed (changes requested):
Hi @vovayatsyuk. The fix totally makes sense for me. Thank you. 

Could I ask you to cover this case with an integration test that will trigger the `InputException`, please? 

Thanks

engcom-Charlie commented:
Hi @vovayatsyuk, could you please cover your changes by integration test?
Thank you.

vovayatsyuk committed:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

vovayatsyuk commented:
There is a test that already checks [account creation with an exception during address save](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/Test/Unit/Model/AccountManagementTest.php#L605-L616). 

The only part that is missing in the test is to check if `delete` wasn't blocked in [beforeDelete](https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Model/AbstractModel.php#L857-L861).

I don't know how to do that since `$this->customerRepository` is a mocked object and no validation is made for this mock.

My zero testing knowledge doesn't allow me to understand how to do this properly. Can you help?

vovayatsyuk commented:
I think I got it. I should add a test to `dev/tests/integration/testsuite/Magento/Customer/Model/AccountManagementTest.php`

vovayatsyuk committed:
Integration test added.
The test checks if email is still available after failed try with incorrect address.

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
Hello, @vovayatsyuk looks like the introduced test falls.
Could you please take a look?
Thank you.

vovayatsyuk commented:
Can you please advice? When I'm running it as a single test is works fine: 

```
../../../vendor/bin/phpunit --filter testCreateAccountWithInvalidAddress testsuite/Magento/Customer/Model/AccountManagementTest.php
```

But when I run whole class:

```
../../../vendor/bin/phpunit testsuite/Magento/Customer/Model/AccountManagementTest.php
```

It fails. Do you have any idea?

engcom-Charlie commented:
@vovayatsyuk you can check the previous test cases in the test class looks like they affect your test case.
Please let me know if you will have any questions.

vovayatsyuk committed:
Improve test: wait for exact exception

vovayatsyuk commented:
Okay. When I run all tests my fix is not working and I still get a `Delete operation is forbidden for current area` exception.

I've updated the test. It still doesn't work, but at least it shows the real issue of why it's failed.

engcom-Charlie commented:
It is necessary to make that all tests passed successfully.
Could you do it?
Thank you.

vovayatsyuk committed:
Added area name to use when running test

vovayatsyuk commented:
I found the reason for the failure (https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/Customer/_files/customer_rollback.php#L24):

```
$registry->register('isSecureArea', false);
```

p.s. There are 398 other test files that use this code too. I believe all of these lines should be removed if test doesn't change it back to unregistered state.

engcom-Charlie commented:
@magento run Integration Tests

vovayatsyuk commented:
I didn't fix it yet.

The easiest way to fix the test is to unregister 'isSecureArea' flag before setting it. But I think it's not correct because **the real problem is registered flags from previous tests**.

However, if you want me to add `->unregister` I'll do that.

engcom-Charlie commented:
> I found the reason for the failure (https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/Customer/_files/customer_rollback.php#L24):
> 
> ```
> $registry->register('isSecureArea', false);
> ```
> 
> p.s. There are 398 other test files that use this code too. I believe all of these lines should be removed if test doesn't change it back to unregistered state.

Could you try to fix it? Maybe create the right fixture and use it instead of the old one whenever it needs.

vovayatsyuk committed:
Do not register 'isSecureArea' flag in the end of test

engcom-Charlie commented:
@magento run Integration Tests

engcom-Charlie commented:
@magento run Integration Tests

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@magento run all tests

vovayatsyuk committed:
Revert "Do not register 'isSecureArea' flag in the end of test"

This reverts commit e0f8f3de4ccec700333d4494f70ddf3c8160af5b.

vovayatsyuk committed:
Unregister 'isSecureArea' in test.
This flag (false) is added in many other fixture files during integration tests.

vovayatsyuk commented:
Since this flag is added in many other fixture files, I've added "unregister" directly into the test.

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@magento run Integration Tests

magento-engcom-team commented:
Hi @rogyar, thank you for the review.
[ENGCOM-8590](https://jira.corp.magento.com/browse/ENGCOM-8590) has been created to process this Pull Request

engcom-Alfa commented:
@magento create issue

engcom-Alfa commented:
Hi @vovayatsyuk . 
Could you add the actual and expected behavior in the PR?
Thanks!

vovayatsyuk commented:
Done.

engcom-Alfa commented:
@vovayatsyuk 

**Manual testing scenario:**

1. Open Magento/Customer/Model/AccountManagement.php and find the following lines:

```php
        try {
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

2. Change them as shown below (Emulate InputException from addressRepository->save method):

```php
        try {
            throw new InputException(__("Test"));                     // <-- this line was added
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

3. Add product to cart as a guest and proceed to checkout.
4. Place an order;
5. Press "Create account" button;
![Screenshot from 2021-01-05 10-56-51](https://user-images.githubusercontent.com/51679138/103636917-59a81680-4f53-11eb-947f-b8817b52f548.png)

6. Fill the form and submit it.

**Before:** 
1. An error "Delete operation is forbidden for current area" is shown.

![Screenshot from 2021-01-05 10-47-09](https://user-images.githubusercontent.com/51679138/103637119-a2f86600-4f53-11eb-91cf-7083e3838b92.png)

2. Customer is created
![Screenshot from 2021-01-05 10-52-09](https://user-images.githubusercontent.com/51679138/103637156-b9062680-4f53-11eb-8b29-0448928106bf.png)

3. Press the "Create account" button again , fill the form and submit it.
Next submits tells that his email is already in use.

![Screenshot from 2021-01-05 12-35-24](https://user-images.githubusercontent.com/51679138/103637289-f074d300-4f53-11eb-86bd-5dc1954054ef.png)


**After:**
1. An error "Test" is shown.
![Screenshot from 2021-01-05 10-42-39](https://user-images.githubusercontent.com/51679138/103637576-5d886880-4f54-11eb-97cb-a956f82f94ba.png)
2. Customer isn't created
3. Press the "Create account" button again , fill the form and submit it.
Next submits show error message "Test".

@vovayatsyuk Could you take a look at behavior before and after the changes?
Thanks!

vovayatsyuk commented:
Yes, the fix works fine. I've updated expected and actual behavior in the first post.

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Open Magento/Customer/Model/AccountManagement.php and find the following lines:

```php
        try {
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

2. Change them as shown below (Emulate InputException from addressRepository->save method):

```php
        try {
            throw new InputException(__("Test"));                     // <-- this line was added
            foreach ($customerAddresses as $address) {
                if (!$this->isAddressAllowedForWebsite($address, $customer->getStoreId())) {
                    continue;
                }
                if ($address->getId()) {
                    $newAddress = clone $address;
                    $newAddress->setId(null);
                    $newAddress->setCustomerId($customer->getId());
                    $this->addressRepository->save($newAddress);
                } else {
                    $address->setCustomerId($customer->getId());
                    $this->addressRepository->save($address);
                }
            }
            $this->customerRegistry->remove($customer->getId());
        } catch (InputException $e) {
            $this->customerRepository->delete($customer);
            throw $e;
        }
```

3. Add product to cart as a guest and proceed to checkout.
4. Place an order;
5. Press "Create account" button;
![Screenshot from 2021-01-05 10-56-51](https://user-images.githubusercontent.com/51679138/103636917-59a81680-4f53-11eb-947f-b8817b52f548.png)

6. Fill the form and submit it.

**Before:** :heavy_multiplication_x: 
1. An error "Delete operation is forbidden for current area" is shown.

![Screenshot from 2021-01-05 10-47-09](https://user-images.githubusercontent.com/51679138/103637119-a2f86600-4f53-11eb-91cf-7083e3838b92.png)

2. Customer is created
![Screenshot from 2021-01-05 10-52-09](https://user-images.githubusercontent.com/51679138/103637156-b9062680-4f53-11eb-8b29-0448928106bf.png)

3. Press the "Create account" button again , fill the form and submit it.
Next submits tells that his email is already in use.

![Screenshot from 2021-01-05 12-35-24](https://user-images.githubusercontent.com/51679138/103637289-f074d300-4f53-11eb-86bd-5dc1954054ef.png)


**After:** :heavy_check_mark: 
1. An error "Test" is shown.
![Screenshot from 2021-01-05 10-42-39](https://user-images.githubusercontent.com/51679138/103637576-5d886880-4f54-11eb-97cb-a956f82f94ba.png)
2. Customer isn't created
3. Press the "Create account" button again , fill the form and submit it.
Next submits show error message "Test".

The functionality of the account creation has also been tested and everything is working properly.

gabrieldagama commented on app/code/Magento/Customer/Model/AccountManagement.php:
```diff
@@ -928,7 +928,9 @@ public function createAccountWithPasswordHash(CustomerInterface $customer, $hash
             }
             $this->customerRegistry->remove($customer->getId());
         } catch (InputException $e) {
+            $this->registry->register('isSecureArea', true, true);
             $this->customerRepository->delete($customer);
+            $this->registry->unregister('isSecureArea');
```
I believe this is not something we are looking to accept, simulation of the secure area here could bring security problems for the platform. Other suggestions to fix this issue would be the usage of transaction and rollback in case some exception has happened?

If the approach above is not suitable, we may need to rethink if the customer deletion here is actually something we are looking for when an exception happens from Address point of view.

vovayatsyuk replied:
Yeah, if we can modify the code to make it possible to rollback customer creation, that would be the proper fix. (We don't want to create a customer if incorrect address data was given)

But I'm not sure if it's possible to rewrite this code to use DB connection directly. Take a look at the [customerRepository::save](https://github.com/magento/magento2/blob/HEAD/app/code/Magento/Customer/Model/ResourceModel/CustomerRepository.php#L200) that is used a few lines above to create a customer.

gabrieldagama replied:
You are right, such changes would have to be done on ResourceModel layer instead, it definitely would be more complicated. Another approach would be adapting the user interface instead, to be honest, logically the Customer does not depend on the address to exist and I believe it shouldn't be deleted if there was a problem on saving it's address, but this approach may also be complicated.

vovayatsyuk replied:
Changing the logic to allow to successfully create an account may be the simplest solution (do not throw an exception from “catch” statement. The only thing that is not possible in this case is to inform the customer about specific error during account creation (invalid address field value).

I will test it tomorrow and let you know about the results.

vovayatsyuk replied:
Works nicely. Here is a step-by-step result:

1. Press "register account" when trying to checkout with multishipping addresses.
2. Fill the form and press Register.
3. Account is successfully created and a user is redirected to multishipping's "Create Shipping Address" step.
4. Save the new address and proceed to the next checkout step.

The only downside is that the customer filled the address on the register account page, but it wasn't saved and no errors were given but I think it's not citical.

gabrieldagama replied:
Hi @vovayatsyuk, good stuff! Can you push the changes? I guess it will be easier for me to try locally!

Thanks!

gabrieldagama reviewed (changes requested):
Hi @vovayatsyuk, thanks for contributing! Can you have a look at my comment below?

Thanks!

vovayatsyuk committed:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

vovayatsyuk committed:
Successfully register customer when invalid address data is recieved during multishipping checkout flow.

Commit:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

gabrieldagama commented:
@magento run all tests

Commit:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

engcom-Charlie committed:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

engcom-Charlie commented:
@magento run all tests

ishakhsuvarov committed:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

ishakhsuvarov commented:
@magento run all tests

ihor-sviziev commented:
Hi @vovayatsyuk,
I see all linked issues are already closed as non-reproducible or already fixed. Could you please update us, does the problem still there?

ihor-sviziev committed:
Merge branch '2.4-develop' into fix-delete-customer-when-address-validation-was-failed

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @vovayatsyuk,
Please provide more context on when this issue might happen. I see all linked issues already closed.

In the past, I had a similar issue on frontend when the customer was automatically created during the checkout process (using a custom checkout).
But then I I analyzed deeper and it seems like address validation can be simply skipped by adding the following flag `ignore_validation_flag` to the customer model:
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/AccountManagement.php#L1649-L1658
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/ResourceModel/Customer/Relation.php#L26-L65
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/ResourceModel/CustomerRepository.php#L265-L285

So having that in mind, it looks like we shouldn't fix this issue because we already have the mechanism for skipping address validation.

ihor-sviziev reviewed (changes requested):
Hi @vovayatsyuk,
Please provide more context on when this issue might happen. I see all linked issues already closed.

In the past, I had a similar issue on frontend when the customer was automatically created during the checkout process (using a custom checkout).
But then I I analyzed deeper and it seems like address validation can be simply skipped by adding the following flag `ignore_validation_flag` to the customer model:
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/AccountManagement.php#L1649-L1658
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/ResourceModel/Customer/Relation.php#L26-L65
https://github.com/magento/magento2/blob/567638657fd8709073a116f965909740de52175c/app/code/Magento/Customer/Model/ResourceModel/CustomerRepository.php#L265-L285

So having that in mind, it looks like we shouldn't fix this issue because we already have the mechanism for skipping address validation.

vovayatsyuk commented:
Ok, let's close this PR
